src/app/upgrade.cpp

```diff
@@ -0,0 +1,351 @@
+        if (!db.open())
+            throw RuntimeError {QCoreApplication::translate("BitTorrent::Session", "Cannot open database. Error: %1").arg(db.lastError().text())};
+
+        // Check if our table exists
+        QSqlQuery query(db);
+        if (query.exec(QString("SELECT name FROM sqlite_master WHERE type='table' AND name='%1'").arg(TABLE_NAME))) {
+            // Unfortunately QSqlQuery::size() doesn't work here
+            if (query.next()) {
+                QString name = query.value(0).toString();
+                if (name == TABLE_NAME)
+                    return;
+            }
+        }
+
+        // Table doesn't exist. Probably 1st run.
+        // Input PRAGMAs and create table
+        query.exec(QLatin1String("PRAGMA auto_vacuum = FULL"));
+        query.exec(QString("CREATE TABLE %1"
+                           "("
+                           "%2 TEXT PRIMARY KEY NOT NULL,"
+                           "%3 BLOB NOT NULL,"
+                           "%4 BLOB,"
+                           "%5 INT DEFAULT -1,"
+                           "%6 INT NOT NULL"
+                           ")"
+                           ).arg(TABLE_NAME, COL_HASH, COL_METADATA, COL_FASTRESUME, COL_QUEUE, COL_VERSION));
+    }
+
+    bool userAcceptsUpgrade()
+    {
+#ifdef DISABLE_GUI
+        printf("\n*** %s ***\n", qUtf8Printable(QObject::tr("Upgrade")));
+        char ret = '\0';
+        do {
+            printf("%s\n"
+                   , qUtf8Printable(QObject::tr("You updated from an older version that saved things differently. You must migrate to the new saving system. You will not be able to use an older version than v4.2.0 again. Continue? [y/n]")));
+            ret = getchar(); // Read pressed key
+        }
+        while ((ret != 'y') && (ret != 'Y') && (ret != 'n') && (ret != 'N'));
+
+        if ((ret == 'y') || (ret == 'Y'))
+            return true;
+#else
+        QMessageBox msgBox;
+        msgBox.setText(QObject::tr("You updated from an older version that saved things differently. You must migrate to the new saving system. If you continue, you will not be able to use an older version than v4.2.0 again."));
+        msgBox.setWindowTitle(QObject::tr("Upgrade"));
+        msgBox.addButton(QMessageBox::Abort);
+        msgBox.addButton(QMessageBox::Ok);
+        msgBox.setDefaultButton(QMessageBox::Abort);
+        msgBox.show(); // Need to be shown or to moveToCenter does not work
+        msgBox.move(Utils::Misc::screenCenter(&msgBox));
+        if (msgBox.exec() == QMessageBox::Ok)
+            return true;
+#endif // DISABLE_GUI
+
+        return false;
+    }
+}
+
+bool upgrade(bool ask)
+{
+    using namespace Utils::DB;
+    const QString backupFolderPath = Utils::Fs::expandPathAbs(specialFolderLocation(SpecialFolder::Data) + "BT_backup");
+    const QDir backupFolderDir(backupFolderPath);
+    const QString dbPath = backupFolderDir.absoluteFilePath(QLatin1String{FILENAME});
+    // If db exists don't try to upgrade
```

Comment:
I think you misunderstood it here.
At the current moment this tries to updrade from the non-db way to the db-way. So if a db exists we don't try to migrate older files (.torrents+.fastresumes). It is assumed that the upgrade was done already and the user dropped new files at a later date.
Unless you suggest that we should always input into the db any kind of .torrent/.fastresume we find during startup.

Reply:
Also in case you missed it: Each row has a `version` column. During startup if the version number is different from the supported one then that row isn't loaded. New versions will upgrade smaller version, while refusing to load bigger version.

Reply:
>I think you misunderstood it here.

Is it? Here's your code: 
```c++
                    // If db exists don't try to upgrade
		    if (QFile::exists(dbPath)) {
		        initiliazeDB(dbPath);
		        return true;
		    }
		    else {
		        if (ask && !userAcceptsUpgrade()) return false;
		    }
```

You ask the user for upgrade if the database doesn't exist. But it is incorrect if the database does not exist due to the fact that the user has never used qBittorrent (on this system).

I've suggested correct logic in another comment.
>Now, I realize that his approach has a drawback. 

Yes, it is.
>Should we instead create a 2nd table named "version" with a single col+row that will hold the version of the whole db?

I think "overall application versioning" I suggested in other comment should fit us. There is no need to have separate versions for each thing.